### ISSUE TITLE
Allow property type to be inferred

### ISSUE THREAD

**Issue by filipkowicz** [square/kotlinpoet#407]
currently i cant find way to use `addProperty(PropertySpec.builder("name", CLASS)` where CLASS is inferred from initializer.

as output I'd like to have
`val name = SOME_OBJECT` where name has type inferred from SOME_OBJECT

**Comment by Egorand** [comment]
What's the use case for this?

**Comment by filipkowicz** [comment]
just to simplify code. Kotlin doesn't require providing types so it could be easier to generate classes tis way.

**Comment by JakeWharton** [comment]
Dupe of #355.

Type inference is an optimization for writing code, not reading it. In this case we only have things to _lose_ by removing the specificity of the generated code.

**Comment by cretz** [comment]
Disappointed here.

> Type inference is an optimization for writing code, not reading it.

That's not true. At the least that's subjective and should be determined by the user of this library. That's like saying you should require emitted code come with comments because you're a fan of them.

**Comment by JakeWharton** [comment]
Ignoring your false equivalence, your counterargument lacks an actual argument. Why is it not true? We're not in the habit of adding foot-guns that reduce the efficacy and accuracy of the code generator without proper justification.

**Comment by JakeWharton** [comment]
Configurability is a detractor. Exposing options is a lazy way to try and please everyone which only winds up reducing the overall API usability. Beyond that, "code quality" isn't defined so I still am not hearing a strong argument for optionality.

It's easy to argue for requiring the type though. For example, without a property type we cannot collapse the combination of a constructor parameter, property declaration, and property initializer into the condensed syntax that everyone expects.
```
class Foo(val bar: Bar)
```

KotlinPoet (and JavaPoet) force you to write with specificity and we automatically optimize the generated output where possible on your behalf. You express the semantics of the structure of the API and not the syntax.

I don't think it's worth worrying about what we would do for lambdas since the library doesn't care about them. We would evaluate the advantages and disadvantages of doing so based on what it afforded us, just like we're doing here.

**Comment by cretz** [comment]
Sure, no prob. I think I might have to look elsewhere. For instance [here](https://github.com/cretz/pb-and-k/blob/202bb73df347ff140f74775813246ef80f884adc/examples/gradle-and-jvm/src/main/kotlin/pbandk/examples/addressbook/pb/addressbook.kt) is an example of some code that a code gen I wrote recently built. I do it by hand w/ strings for configurability and because I wanted the generator to be multiplatform. You can see a lot of inference there to keep it what I would consider readable and this lib couldn't generate such code, but my needs for that project and the next project I'm working on (a transpiler) just have too many formatting requirements to use this opinionated lib. I will say I think if it's reasonable for a human to type it, and it's reasonable for a lint to require it, it should be reasonable to generate it.

**Comment by swankjesse** [comment]
Yeah if you need that much control this is not the library for you.

Eliding types in general is a difficult problem. Consider writing a reformatter to elide unnecessary types from this program:

```
fun a(): String = b()
fun b(): String = a()
```

We don’t want developers to spend their time deciding which types to elide.

**Comment by cretz** [comment]
@swankjesse that's a recursive compiler error to remove both return types. Regardless, I got the hint here about inflexibility and solved my problems with https://github.com/cretz/kastree. Libraries like this one will never pass human lint/syntax consistency rules such as disallowing explicit types on const vals.

**Comment by JakeWharton** [comment]
That was his point, yes

**Comment by cHullaert** [comment]
I have this part of code 
``
    private val friends = mutableListOf<Person>()
``
with inference its possible to use all the interface implemented List and MutableCollection, here I've a cast mandatory since I can have only one of this types
